SofaKernel/framework/sofa/core/DataEngine.h

```diff
@@ -184,6 +184,66 @@ class SOFA_CORE_API DataEngine : public core::DataTrackerDDGNode, public virtual
 
 };
 
+/**
+ *  \brief A simpler API for your DataEngines
+ *
+ * Implementation good rules:
+ *
+ * void init()
+ * {
+ *    addInput // indicate all inputs
+ *    addOutput // indicate all outputs
+ *    setDirtyValue(); // You still have to set that manually, sorry
+ * }
+ *
+ * // optional (called each time a data is modified in the gui)
+ * // it is not always desired
+ * void reinit()
+ * {
+ *    update();
+ * }
+ *
+ * void doUpdate() override
+ * {
+ *    don't overthink it, just code. acces your inputs, set your outputs
+ *    No cleanDirty() required
+ * }
+ *
+ */
+class SOFA_CORE_API SimpleDataEngine : public sofa::core::DataEngine
+{
+ public:
+  SOFA_CLASS(SimpleDataEngine, sofa::core::DataEngine);
+
+  SimpleDataEngine() : Inherit1() {}
+  virtual ~SimpleDataEngine() {}
+
+  /// Updates your inputs and calls cleanDirty() for you.
+  /// User implementation moved to doUpdate()
+  virtual void update() final
+  {
+      for(auto& input : getInputs())
+      {
+          static_cast<sofa::core::objectmodel::BaseData*>(input)
+                  ->updateIfDirty();
+      }
+      cleanDirty();
+      doUpdate();
+  }
+
+  /// Where you put your engine's impl
+  virtual void doUpdate() = 0;
+
+  /// Automatically adds the input fields to the datatracker
+  void addInput(sofa::core::objectmodel::BaseData* data)
+  {
+      m_dataTracker.trackData(*data);
```

Comment:
I'm not use to the dataTracker, if the data are added in the dataTracker, is it not possible to use it in the update method to check the data instead of calling the updateIfDirty?

Reply:
The datatracker is not influenced by the updateIfDirty() call. You can absolutely use the datatracker in the update() method if needed (to check which inputs have been changed, and to specific actions for each dirty data. The call to UpdateIfDirty() simply guarantees that all data have been accessed (thus updated upstream) before calling cleanDirty().

Reply:
ok so the dataTracker just return dirty if on of its Data is dirty?

Reply:
yes, the dataTracker is not at all used in the internal mechanism of the DataEngine class actually. It's a helper object (one instance per dataEngine) that tracks data. It has a very simple interface, and you can just call
` trackData(BaseData*)` << sets a datafield in its map. for now on, its counter will be tracked
` clean() / clean(BaseData*)` << resets the data's counter 
` isDirty() : isDirty()` << checks if the counter is up to date

You can then use this in your update() ( or doUpdate() now) method to see if a specific datafield is dirty or not.